File: apps/node/test/fizzbuzz/domain/model/fizz-buzz-list.test.ts
import { describe, expect, test } from "vitest";
import { FizzBuzzList, FizzBuzzValue } from "../../../../src/fizzbuzz/index";

describe("FizzBuzzList", () => {
  test("空リストを生成できる", () => {
    const list = new FizzBuzzList();
    expect(list.size).toBe(0);
  });

  test("add で新しいリストを返す（不変）", () => {
    const list = new FizzBuzzList();
    const newList = list.add(new FizzBuzzValue("1", 1));
    expect(list.size).toBe(0);
    expect(newList.size).toBe(1);
  });

  test("toStringArray で文字列配列を返す", () => {
    let list = new FizzBuzzList();
    list = list.add(new FizzBuzzValue("1", 1));
    list = list.add(new FizzBuzzValue("2", 2));
    list = list.add(new FizzBuzzValue("Fizz", 3));
    expect(list.toStringArray()).toEqual(["1", "2", "Fizz"]);
  });

  test("イテレータで反復できる", () => {
    let list = new FizzBuzzList();
    list = list.add(new FizzBuzzValue("Fizz", 3));
    list = list.add(new FizzBuzzValue("Buzz", 5));

    const values: string[] = [];
    for (const v of list) {
      values.push(v.toString());
    }
    expect(values).toEqual(["Fizz", "Buzz"]);
  });
});

File: apps/ruby/test/fizz_buzz/domain/model/fizz_buzz_list_test.rb
# frozen_string_literal: true

require_relative '../../../test_helper'
require_relative '../../../../lib/fizz_buzz/fizz_buzz'

class FizzBuzzListTest < Minitest::Test
  def test_空リストを生成できる
    list = FizzBuzzList.new
    assert_equal 0, list.size
  end

  def test_addで新しいリストを返す
    list = FizzBuzzList.new
    values = [FizzBuzzValue.new('1', 1)]
    new_list = list.add(values)
    assert_equal 0, list.size
    assert_equal 1, new_list.size
  end

  def test_to_string_arrayで文字列配列を返す
    values = [
      FizzBuzzValue.new('1', 1),
      FizzBuzzValue.new('2', 2),
      FizzBuzzValue.new('Fizz', 3)
    ]
    list = FizzBuzzList.new(values)
    assert_equal %w[1 2 Fizz], list.to_string_array
  end

  def test_上限を超えるとエラー
    values = (1..101).map { |number| FizzBuzzValue.new(number.to_s, number) }
    assert_raises RuntimeError do
      FizzBuzzList.new(values)
    end
  end
end
